sympy/ntheory/factor_.py

```diff
@@ -1009,13 +1009,10 @@ def factorint(n, limit=None, use_trial=True, use_rho=True, use_pm1=True,
         x = n.args[0]
         factors = {}
         for p in sieve.primerange(2, x):
-            m = 0
-            d = p
-            q = x // p
+            m, q = 0, x // p
```

Comment:
Could the loop be reorganized to make use of the factors that have been found to reduce the size of `x`? After the first run with p = 2, one could continue with `x // 2**m` instead of `x`, etc.

Reply:
Yes, `x` should not be reduced in this case. (The connection with factorial is not visible in github diff. It only says `factorint`).

Reply:
That little loop is pretty clever. Would the following mod be worthwhile?
```
m = 2  # get first pass started
for p in sieve.primerange(2, x):
    if m > 1:
        m, q = 0, x // p
        while q:
            m += q
            q //= p
    factors[p] = m
```

Reply:
This will make `factors[2] = 2` always, isnt it.

Reply:
```
-    # for unevaluated factorial, if n < 20!, direct computation is faster
-    # since it uses lookup table

if isinstance(n, factorial):
    x = as_int(n.args[0])
    if x >= 20:
        factors = {}
        m = 2  # to initialize `if` below
        for p in sieve.primerange(2, x):
            if m > 1:
                m = 0
                q = x // p
                while q != 0:
                    m += q
                    q = x // d
            factors[p] = m
        if factors and verbose:
            for k in sorted(factors):
                print(factor_msg % (k, factors[k]))
        if verbose:
            print(complete_msg)
        return factors
    else:
        # if n < 20!, direct computation is faster
        # since it uses a lookup table
        n = n.func(x)  # evaluate and continue below

>>> factorint(factorial(1, evaluate=0))
{}
>>> factorint(factorial(2, evaluate=0))
{2: 1}
>>> factorint(factorial(3, evaluate=0))
{2: 1, 3: 1}
>>> factorint(factorial(4, evaluate=0))
{2: 3, 3: 1}
```

At least unevaluate 1! should be added to the test suite.

Reply:
My last comment was not correct.
```
m = 2  # to initialize `if` below
for p in sieve.primerange(2, x):
    if m > 1:
        m, q = 0, x // p
        while q:
            m += q
            q //= p
    factors[p] = m
```
I get the trick, so basically this directly assigns primes with exponent `1`. Nice.

What does `n.func(x)` mean?

Reply:
Figured out `n.func(x)`. Made the changes.
Thank you for the review.